qiskit/synthesis/discrete_basis/commutator_decompose.py

```diff
@@ -53,16 +53,42 @@ def _compute_rotation_axis(matrix: np.ndarray) -> np.ndarray:
     """
     _check_is_so3(matrix)
 
+    # If theta represents the rotation angle, then trace = 1 + 2cos(theta).
     trace = _compute_trace_so3(matrix)
-    theta = math.acos(0.5 * (trace - 1))
-    if math.sin(theta) > 1e-10:
-        x = 1 / (2 * math.sin(theta)) * (matrix[2][1] - matrix[1][2])
-        y = 1 / (2 * math.sin(theta)) * (matrix[0][2] - matrix[2][0])
-        z = 1 / (2 * math.sin(theta)) * (matrix[1][0] - matrix[0][1])
-    else:
+
+    if trace >= 3 - 1e-10:
+        # The matrix is the identity (rotation by 0)
         x = 1.0
         y = 0.0
         z = 0.0
+
+    elif trace <= -1 + 1e-10:
+        # The matrix is the 180-degree rotation
+        # We have:
+        #   k_x^2 = (1 + r_00) / 2, k_y^2 = (1 + r_11) / 2, k_z^2 = (1 + r_22) / 2,
+        #   k_x k_y = r_01 / 2, k_y k_z = r_12 / 2, k_x k_z = r_02 / 2
+        squares = [0.5 * (1 + matrix[0][0]), 0.5 * (1 + matrix[1][1]), 0.5 * (1 + matrix[2][2])]
+        index_of_max = squares.index(max(squares))
+        if index_of_max == 0:
+            x = math.sqrt(squares[0])
+            y = matrix[0][1] / (2 * x)
+            z = matrix[0][2] / (2 * x)
+        elif index_of_max == 1:
+            y = math.sqrt(squares[1])
+            x = matrix[0][1] / (2 * y)
+            z = matrix[1][2] / (2 * y)
+        else:
+            z = math.sqrt(squares[2])
+            x = matrix[0][2] / (2 * z)
+            y = matrix[1][2] / (2 * z)
+
+    else:
+        # The matrix is the rotation by theta with sin(theta)!=0
+        theta = math.acos(0.5 * (trace - 1))
+        x = 1 / (2 * math.sin(theta)) * (matrix[2][1] - matrix[1][2])
```

Comment:
We should use the antisymmetrized matrix here, right? I.e. `r = (matrix - matrix.T) / 2` instead of `matrix`

Reply:
This is how it was before my changes, so (your argument on minimizing changes) let's not change it. But yes, ``matrix[2][1] - matrix[1][2]`` is exactly an entry in ``matrix - matrix.T``. 

(Note: as the sign is important, the entry to choose is the entry ``matrix[2][1] - matrix[1][2]`` and not ``matrix[1][2] - matrix[2][1]``, as follows from the calculation based on Rodrigues rotation formula).

Reply:
I mean that this should be
```python
r = (matrix - matrix.T) / 2
x = 1 / (2 * math.sin(theta)) * (r[2][1] - r[1][2])
y = # ...
```
no? 🙂 I think this was wrong in the original code.